doc/solver.md

```diff
@@ -0,0 +1,710 @@
+| 8 | `foo 1.0.0` | decision | decision making | | 1 |
+| 9 | `bar 1.1.0` | decision | decision making | | 2 |
+
+In step 6, the decision making process considers `foo 1.1.0` by adding its
+dependency as the incompatibility `{foo >=1.1.0, not bar ^2.0.0}`. However, if
+`foo 1.1.0` were selected, this incompatibility would be satisfied: `foo 1.1.0`
+satisfies `foo >=1.1.0`, and `bar ^1.0.0` from step 5 satisfies `not bar
+^2.0.0`. So decision making ends without selecting a version, and unit
+propagation is run again.
+
+Unit propagation determines that the new incompatibility, `{foo >=1.1.0, not bar
+^2.0.0}`, is almost satisfied (again because `bar ^1.0.0` satisfies `not bar
+^2.0.0`). Thus it's able to deduce `not foo >=1.1.0` in step 7, which lets the
+next iteration of decision making choose `foo 1.0.0` which is compatible with
+`root`'s constraint on `bar`.
+
+## Performing Conflict Resolution
+
+This example shows full conflict resolution in action. Given the following
+packages:
+
+* `root 1.0.0` depends on `foo >=1.0.0`.
+* `foo 2.0.0` depends on `bar ^1.0.0`.
+* `foo 1.0.0` has no dependencies.
+* `bar 1.0.0` depends on `foo ^1.0.0`.
+
+Pubgrub goes through the following steps:
+
+| Step | Value | Type | Where it was added | Cause | Decision level |
+| ---- | ----- | ---- | ------------------ | ----- | -------------- |
+| 1 | `root 1.0.0` | decision | top level | | 0 |
+| 2 | `{root 1.0.0, not foo >=1.0.0}` | incompatibility | top level | | |
+| 3 | `foo >=1.0.0` | derivation | unit propagation | step 2 | 0 |
+| 4 | `{foo >=2.0.0, not bar ^1.0.0}` | incompatibility | decision making | | |
+| 5 | `foo 2.0.0` | decision | decision making | | 1 |
+| 6 | `bar ^1.0.0` | derivation | unit propagation | step 4 | 1 |
+| 7 | `{bar any, not foo ^1.0.0}` | incompatibility | decision making | | |
+
+The incompatibility added at step 7 is satisfied by the partial assignment: `bar
+any` is satisfied by `bar ^1.0.0` from step 6, and `not foo ^1.0.0` is satisfied
+by `foo 2.0.0` from step 5. This causes Pubgrub to enter conflict resolution,
+where it iteratively works towards the root cause of the conflict:
+
+| Step | Incompatibility | Term | Satisfier | Cause | Previous Satisfier |
+| ---- | --------------- | ---- | --------- | ----- | ------------------ |
+| 8 | `{bar any, not foo ^1.0.0}` | `bar any` | `bar ^1.0.0` from step 6 | `{foo >=2.0.0, not bar ^1.0.0}` | `foo 2.0.0` from step 5 |
+| 9 | `{foo >=2.0.0}` | `foo >=1.0.0` | `foo 2.0.0` from step 5 | | |
+
+In step 9, we merge the two incompatibilities `{bar any, not foo ^1.0.0}` and
+`{foo >=2.0.0, not bar ^1.0.0}` as described in
+[conflict resolution](#conflict-resolution), to produce `{not foo ^1.0.0,
+foo >=2.0.0, bar any ∪ not bar ^1.0.0}`. Since `not not bar ^1.0.0 = bar ^1.0.0`
+satisfies `bar any`, this simplifies to `{not foo ^1.0.0, foo >=2.0.0}`. We can
+then take the intersection of the two `foo` terms to get `{foo >=2.0.0}`.
```

Comment:
This paragraph is strangely formatted (see https://github.com/dart-lang/pub/blob/e891bfbfaaddf07b7906abeb424cda8c8efd8579/doc/solver.md#performing-conflict-resolution)

Reply:
Done.